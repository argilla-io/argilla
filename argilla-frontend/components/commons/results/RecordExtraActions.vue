<!--
  - coding=utf-8
  - Copyright 2021-present, the Recognai S.L. team.
  -
  - Licensed under the Apache License, Version 2.0 (the "License");
  - you may not use this file except in compliance with the License.
  - You may obtain a copy of the License at
  -
  -     http://www.apache.org/licenses/LICENSE-2.0
  -
  - Unless required by applicable law or agreed to in writing, software
  - distributed under the License is distributed on an "AS IS" BASIS,
  - WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  - See the License for the specific language governing permissions and
  - limitations under the License.
  -->

<template>
  <div v-click-outside="close" :key="open" class="record__extra-actions">
    <base-button class="small clear" @click.prevent="open = !open"
      ><svgicon name="meatballs" width="14" color="#4A4A4A" />
    </base-button>
    <div v-if="open" class="extra-actions__content">
      <div @click="showRecordInfoModal()">
        <span>View record info</span>
      </div>
      <base-action-tooltip tooltip="Copied">
        <div @click="$copyToClipboard(recordClipboardText)">
          <span>Copy text</span>
        </div>
      </base-action-tooltip>
    </div>
  </div>
</template>

<script>
import "assets/icons/meatballs";
import { IdState } from "vue-virtual-scroller";

export default {
  mixins: [
    IdState({
      // You can customize this
      idProp: (vm) => `${vm.datasetName}-${vm.recordId}`,
    }),
  ],
  props: {
    recordId: {
      type: String | Number,
      required: true,
    },
    recordClipboardText: {
      type: Array | String,
      required: true,
    },
    datasetName: {
      type: String,
      required: true,
    },
  },
  idState() {
    return {
      open: false,
    };
  },
  computed: {
    open: {
      get() {
        return this.idState.open;
      },
      set(newValue) {
        this.idState.open = newValue;
      },
    },
  },
  beforeDestroy() {
    this.close();
  },
  methods: {
    showRecordInfoModal() {
      this.$emit("show-record-info-modal");
      this.close();
    },
    close() {
      this.open = false;
    },
  },
};
</script>

<style lang="scss" scoped>
.extra-actions {
  position: relative;
  &__content {
    position: absolute;
    right: 0.7em;
    top: calc(100% + $base-space);
    background: white;
    border-radius: $border-radius;
    box-shadow: $shadow;
    padding: 3px;
    min-width: 135px;
    z-index: 1;
    .disabled {
      pointer-events: none;
      opacity: 0.7;
    }
    div {
      padding: 0.5em;
      color: $black-54;
      cursor: pointer;
      display: block;
      background: white;
      transition: background 0.3s ease-in-out;
      &:first-child {
        border-top-left-radius: $border-radius;
        border-top-right-radius: $border-radius;
      }
      &:last-child {
        border-bottom-left-radius: $border-radius;
        border-bottom-right-radius: $border-radius;
      }
      &:hover {
        transition: background 0.3s ease-in-out;
        background: $black-4;
      }
    }
  }
}
</style>
