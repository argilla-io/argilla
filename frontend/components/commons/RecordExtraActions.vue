<template>
  <div class="record__extra-actions">
    <div v-if="hasMetadata" @click="$emit('onShowMetadata')">
      <span>View metadata</span>
    </div>

    <div
      v-for="status in allowedStatusActions"
      :key="status.key"
      @click="onChangeRecordStatus(status.key)"
    >
      <span>{{ status.name }}</span>
    </div>
  </div>
</template>

<script>
import { BaseRecord } from "@/models/Common";

export default {
  props: {
    annotationMode: {
      type: Boolean,
      default: false,
    },
    record: {
      type: BaseRecord,
      required: true,
    },
  },
  data: () => ({
    statusActions: [
      {
        name: "Discard",
        key: "Discarded",
        class: "discard",
      },
    ],
  }),
  computed: {
    hasMetadata() {
      const metadata = this.record.metadata;
      return metadata && Object.values(metadata).length;
    },
    recordStatus() {
      return this.record.status;
    },
    allowedStatusActions() {
      if (!this.annotationMode) {
        return [];
      }

      return this.statusActions.map((status) => ({
        ...status,
        isActive: this.recordStatus === status.key,
      }));
    },
  },
  methods: {
    onChangeRecordStatus(status) {
      if (this.record.status !== status) {
        this.$emit("onChangeRecordStatus", status);
      } else {
        this.$emit("onChangeRecordStatus", 'Edited');
      }
    },
  },
};
</script>

<style lang="scss" scoped>
.record {
  &__extra-actions {
    line-height: 1;
    text-align: left;
    color: $font-secondary;
    margin-top: 1em;
    margin-bottom: 1em;
    @include font-size(13px);
    .annotate {
      color: $success;
    }
    .discard {
      color: $error;
    }
    > div {
      margin-top: 0;
    }
    > * + *:before {
      content: "";
      margin: auto 1em;
      height: 1em;
      width: 1px;
      background: $font-medium-color;
      vertical-align: middle;
      display: inline-block;
    }
    & > * {
      display: inline-block;
      cursor: pointer;
    }
  }
}
</style>
