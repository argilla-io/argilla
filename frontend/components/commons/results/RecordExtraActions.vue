<!--
  - coding=utf-8
  - Copyright 2021-present, the Recognai S.L. team.
  -
  - Licensed under the Apache License, Version 2.0 (the "License");
  - you may not use this file except in compliance with the License.
  - You may obtain a copy of the License at
  -
  -     http://www.apache.org/licenses/LICENSE-2.0
  -
  - Unless required by applicable law or agreed to in writing, software
  - distributed under the License is distributed on an "AS IS" BASIS,
  - WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  - See the License for the specific language governing permissions and
  - limitations under the License.
  -->

<template>
  <div class="record__extra-actions">
    <div v-if="hasMetadata" @click="$emit('onShowMetadata')">
      <span>View metadata</span>
    </div>
    <template v-if="allowChangeStatus">
      <div
        v-for="status in allowedStatusActions"
        :key="status.key"
        @click="onChangeRecordStatus(status.key)"
      >
        <span>{{ status.name }}</span>
      </div>
    </template>
  </div>
</template>

<script>
import { BaseRecord } from "@/models/Common";

export default {
  props: {
    allowChangeStatus: {
      type: Boolean,
      default: false,
    },
    record: {
      type: BaseRecord,
      required: true,
    },
  },
  data: () => ({
    statusActions: [
      {
        name: "Discard",
        key: "Discarded",
        class: "discard",
      },
    ],
  }),
  computed: {
    hasMetadata() {
      const metadata = this.record.metadata;
      return metadata && Object.values(metadata).length;
    },
    recordStatus() {
      return this.record.status;
    },
    allowedStatusActions() {
      return this.statusActions.map((status) => ({
        ...status,
        isActive: this.recordStatus === status.key,
      }));
    },
  },
  methods: {
    onChangeRecordStatus(status) {
      if (this.record.status !== status) {
        this.$emit("onChangeRecordStatus", status);
      } else {
        this.$emit("onChangeRecordStatus", "Edited");
      }
    },
  },
};
</script>

<style lang="scss" scoped>
.record {
  &__extra-actions {
    line-height: 1;
    text-align: left;
    color: $font-secondary;
    margin-top: 0.5em;
    margin-bottom: 1.5em;
    @include font-size(13px);
    padding-left: 2.3em;
    .list__item--annotation-mode & {
      padding-left: 65px;
    }
    .annotate {
      color: $success;
    }
    .discard {
      color: $error;
    }
    > div {
      margin-top: 0;
    }
    > * + *:before {
      content: "";
      margin: auto 1em;
      height: 1em;
      width: 1px;
      background: $font-medium-color;
      vertical-align: middle;
      display: inline-block;
    }
    & > * {
      display: inline-block;
      cursor: pointer;
    }
  }
}
</style>
