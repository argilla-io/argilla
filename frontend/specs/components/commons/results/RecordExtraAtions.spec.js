import { mount } from "@vue/test-utils";
import RecordExtraActions from "@/components/commons/results/RecordExtraActions";

function mountRecordExtraActions() {
  return mount(RecordExtraActions, {
    idState() {
      return {
        open: true,
      };
    },
    stubs: ["base-action-tooltip"],
    propsData: {
      recordId: "recordId",
      recordClipboardText: ["word 1", "word 2"],
      datasetName: "datasetName",
      record: {
        metadata: {
          Metadata: "metadata example",
        },
        status: "Validated",
      },
      allowChangeStatus: true,
      dataset: {
        name: "Dataset1",
      },
    },
  });
}

describe("RecordExtraActions", () => {
  let spy = jest.spyOn(console, "error");
  afterEach(() => spy.mockReset());

  test("renders properly", () => {
    const wrapper = mountRecordExtraActions();
    expect(wrapper.html()).toMatchSnapshot();
  });
});
