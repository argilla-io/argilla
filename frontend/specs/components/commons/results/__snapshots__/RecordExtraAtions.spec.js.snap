// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RecordExtraActions renders properly 1`] = `
<div class="record__extra-actions" record="[object Object]" allowchangestatus="true" dataset="[object Object]">
  <base-button-stub class="small clear"><svg version="1.1" viewBox="0 0 30 8" class="svg-icon svg-fill" style="width: 14px;">
      <path fill="#4A4A4A" stroke="none" pid="0" d="M7.5 4A3.75 3.75 0 110 4a3.75 3.75 0 017.5 0zM18.75 4a3.75 3.75 0 11-7.5 0 3.75 3.75 0 017.5 0zM26.25 7.75a3.75 3.75 0 100-7.5 3.75 3.75 0 000 7.5z" _fill="#000"></path>
    </svg></base-button-stub>
  <div class="extra-actions__content">
    <div><span>View record info</span></div>
    <base-action-tooltip-stub tooltip="Copied">
      <div><span>Copy text</span></div>
    </base-action-tooltip-stub>
  </div>
</div>
`;
